File: src/sort/quickSortLomuto.ts
import {Sort} from "./Sort";
export default class QuickSortLomuto extends Sort{
    sort(array: number[]): void | number[] {
        this.divide(array,0,array.length-1);
    }
    divide(array:number[],low:number,high:number):void{
        if(low >= high)return;
        let p = this.partition(array,low,high);
        this.divide(array,0,p-1);
        this.divide(array,p+1,high);
    }
    private partition(array:number[],low:number,high:number): number {
        const pivot = array[high];
        let i = low;
        for(let j=low;j<high;j++){
            if(array[j] <= pivot){
                QuickSortLomuto.swap(array,i,j);
                i++;
            }
        }
        QuickSortLomuto.swap(array,i,high);
        return i;
    }
}

File: sortinplace/QuickSortLomuto.java
package sortinplace;

public class QuickSortLomuto extends QuickSort {

    public QuickSortLomuto(int[] array,int low,int high){
        super(array,low,high);
    }
    @Override
    protected void sort(){
        if(this.low >= this.high) return;
        int p = this.partition();
        invokeAll(new QuickSortLomuto(array,this.low,p-1),
                new QuickSortLomuto(array,p+1,this.high));
    }
    @Override
    protected int partition(){
        int pivot = this.array[this.high];
        int i = this.low;
        for(int j = this.low;j<this.high;j++){
            if(this.array[j] <= pivot){
                ISwapper.swap(this.array,j,i);
                i++;
            }
        }
        ISwapper.swap(this.array,this.high,i);
        return i;
    }
}